### ISSUE TITLE
Background_sub should do sigma-clipping when computing average

### ISSUE THREAD

**Issue by hbushouse** [spacetelescope/jwst#1078]
The background subtraction step used in the calspec2 (and eventually calimage2) pipeline to subtract dedicated background exposures from science target exposures currently produces a combined background image (when there are multiple background inputs) by computing a simple mean of the inputs. The JCCWG would like to also have sigma-clipping applied, in an attempt to remove any sources that might be present in some of the input images.

**Comment by hbushouse** [comment]
@karllark @drlaw1558 The CALWG description indicates:
```
The combined background image will be constructed by averaging the data from all background exposures in detector space and will use a sigma-clipping algorithm to remove outliers.
```
Any thoughts/desires as to the exact details of this? Should we in fact use a straight average/mean as opposed to, say, median? Should the choice of mean vs median vary as a function of the number of images being combined? Should there be a single iteration of sigma-clipping or multiple? What should the default sigma value be?

**Comment by philhodge** [comment]
Here's an alternative that looks good to me.

    limits = np.percentile(x, (lowlim, highlim))
    mask = np.logical_and(x >= limits[0], x <= limits[1])
    mean_value = x[mask].mean(dtype=np.float)

`lowlim` and `highlim` are in percent, e.g. 25 and 75 respectively.
Sigma clipping would do a poor job for an array such as this:

    [1, 3, 5, 7, 9, 20, 40, 60, 80, 91, 93, 95, 97, 99]

**Comment by hbushouse** [comment]
Interesting. I guess it depends a lot on what kinds of variation we're expecting within a given pixel for a series of background images. If it's the case that the actual background signal does not vary all that much from image to image, such that we're mainly interested in clipping out sources, and those sources only show up in a given pixel in a small number (1 or 2) of images because of dithering, then I would think we'd usually be dealing with fairly normal looking (not in the official statistical sense of "normal") distribution of values containing a small number of outliers. E.g. something more like:
```
[2, 5, 42, 3, 4, 6, 3, 50]
```
But even in this type of case we might be dealing with some values that are the low-level wings of a source that's sitting a couple pixels away, and regular sigma clipping is not going to be very good at picking out those values that are elevated only a small amount. That's where something like a percentage clip would be useful (e.g. always clip out anything that's >80%).

**Comment by philhodge** [comment]
I think sigma clipping is probably best for realistic data, but I would rather start with the median instead of the mean.  And instead of setting sigma to the standard deviation of all data, you could use the median of the absolute values of deviations (from the median).  That is, try to get rid of the wild outliers immediately.

**Comment by karllark** [comment]
The sigma clipping that I am used to computes the mean and standard deviation and then removes any points that are n sigma away from the mean.  Where n is usually something like 5.  I agree that this will be hard with a small number of points, but any such clipping algorithm will have problems with low numbers of points.

While we could use the uncertainty propagated through the pipeline, this is not a good idea.  The pipeline provided uncertainty does not include variations due to the source - e.g., a strong emission line or a point source (for imaging). 

One question, what sigma clipping algorithm are we using when we combine imaging observations?  The goal for such clipping is to remove cosmic rays that got past the ramp based CR detection.

**Comment by hbushouse** [comment]
Yes, some available routines, like astropy.stats.sigma_clip, allow you to select the "center function" that's used for computing the standard deviation and defaults to median, even though the final result is giving you the mean. That way the computation of stdev isn't already biased by 1 or 2 outliers.

**Comment by karllark** [comment]
Median is fine for determining the point to use for comparison for clipping.  As you just said @hbushouse : the final value should be based on a mean instead of median.

**Comment by drlaw1558** [comment]
If I were to use IDL-speak, I'd say that we want https://svn.sdss.org/public/repo/manga/mangadrp/tags/v2_1_2/pro/math/ml_meanclip.pro
I.e., I'm agreeing that the intial clipping is done about the median, but the mean is what is eventually returned.  The routine should be iterated to convergence (not just a single clip)- does the astropy algorithm do this?  It's also useful to be able to specify values to be ignored, and to think about the performance in the realm of 2,3, and 4 data points.

**Comment by larrybradley** [comment]
Yes, `astropy.stats.sigma_clip` can clip to convergence by setting `maxiters=None` (dev version) or  `iters` (now deprecated) in released versions.

**Comment by hbushouse** [comment]
`astropy.stats.sigma_clip` has many such options that allow flexibility, including:
- sigma: defaults to 3, for symmetric lower and upper limits
- sigma_lower: defaults to None (use symmetric sigma limits)
- sigma_upper: defaults to None (use symmetric sigma limits)
- iters: number of iterations or iterate to convergence if None (default = 5)
- cenfunc: function used to compute clipping center (defaults to numpy.median)
- stdfunc: function used to compute standard deviation (defaults to numpy.std)

Note that the result of this function is a new copy of the array with clipped values masked out. You can then choose any function for computing the final result, e.g. mean of the non-masked values.

The ability to ignore specified values could be implemented separately by just adding to the masking determined by sigma_clip. We could also wrap desired behavior for small numbers of data points.

**Comment by hbushouse** [comment]
Basic functionality implemented in #2270. Any future updates will be handled in new tickets.
